aten/src/ATen/native/cuda/BatchLinearAlgebra.cu

```diff
@@ -2799,14 +2800,107 @@ static void apply_gels(const Tensor& a, Tensor& b, Tensor& infos) {
+  auto m = A.size(-2);
+  auto n = A.size(-1);
+  auto mn = std::min(m, n);
+
+  // explicitly broadcast the batch dimensions of A
+  // TODO: revisit this later to use batch_iterator_with_broadcasting in triangular_solve
+  IntArrayRef A_batch_sizes(A.sizes().data(), A.dim() - 2);
+  IntArrayRef B_batch_sizes(B.sizes().data(), B.dim() - 2);
+  std::vector<int64_t> expand_batch_portion = at::infer_size(A_batch_sizes, B_batch_sizes);
+
+  auto tau_shape = A.sizes().vec();
+  tau_shape.pop_back();
+  tau_shape.back() = mn;
+  Tensor tau = at::empty(tau_shape, A.options());
+
+  if (m >= n) {
+    // Step 1: compute QR factorization using geqrf
+    geqrf_kernel(A, tau);
+
+    // explicitly broadcast the batch dimensions of A
+    // we do it after geqrf so that we don't do redundant computations for the same input
+    auto A_expand_batch = expand_batch_portion;
+    A_expand_batch.insert(A_expand_batch.end(), {A.size(-2), A.size(-1)});
+    Tensor A_expanded = A.expand({A_expand_batch});
+    bool is_fortran_contiguous = A_expanded.transpose(-2, -1).is_contiguous();
+    Tensor A_broadcasted = is_fortran_contiguous ? A_expanded : cloneBatchedColumnMajor(A_expanded);
+    auto tau_expand_batch = expand_batch_portion;
+    tau_expand_batch.push_back(tau.size(-1));
+    Tensor tau_broadcasted = tau.expand({tau_expand_batch}).contiguous();
+
+    // Step 2: B <- Q^H B
+    ormqr_kernel(A_broadcasted, tau_broadcasted, B, /*left=*/true, /*transpose=*/true);
+
+    // Step 3: solve R X = B
+    bool upper = true;
+    bool transpose = false;
+    bool conjugate_transpose = false;
+    bool unitriangular = false;
+    triangular_solve_kernel(
+        const_cast<Tensor&>(A_broadcasted),
+        const_cast<Tensor&>(B),
+        const_cast<Tensor&>(infos),
+        upper, transpose, conjugate_transpose, unitriangular);
+  } else { // underdetermined case
+    Tensor Ah = cloneBatchedColumnMajor(A.conj().transpose(-2, -1));
+
+    // Step 1: compute QR factorization of conjugate transpose of A using geqrf
+    geqrf_kernel(Ah, tau);
+
+    // explicitly broadcast the batch dimensions of A
+    // we do it after geqrf so that we don't do redundant computations for the same input
+    auto A_expand_batch = expand_batch_portion;
+    A_expand_batch.insert(A_expand_batch.end(), {Ah.size(-2), Ah.size(-1)});
+    Tensor Ah_expanded = Ah.expand({A_expand_batch});
+    bool is_fortran_contiguous = Ah_expanded.transpose(-2, -1).is_contiguous();
+    Tensor Ah_broadcasted = is_fortran_contiguous ? Ah_expanded : cloneBatchedColumnMajor(Ah_expanded);
+
+    // Step 2: R^H Z = B
+    bool upper = true;
+    bool transpose = true;
+    bool conjugate_transpose = true;
+    bool unitriangular = false;
+    triangular_solve_kernel(
+        const_cast<Tensor&>(Ah_broadcasted),
+        const_cast<Tensor&>(B),
+        const_cast<Tensor&>(infos),
+        upper, transpose, conjugate_transpose, unitriangular);
+
+    B.narrow(-2, m, n - m).zero_();
```

Comment:
This is because `triangular_solve_kernel` writes its output into the first `m` elements of `B`, right? Could you leave a comment explaining this here?